### DISCUSSION TITLE
node-types.json for generating typed wrappers

### DISCUSSION THREAD

**Discussion by JoranHonig** [tree-sitter/tree-sitter#2095] [Q&A]
I'm building something to generate a typed wrapper around tree-sitter parsers. 

I encountered two issues that might be worth slight adaptions to fix.

### Named extra
A comment might appear anywhere in the language, and would be put in the extras array.

It seems impossible to recognize a type as "extra", with just the node types.

Having a boolean like `is_extra` on every named type would help generating typed bindings.

### Unnamed tokens
`node-types.json` has a range of unnamed types in there.

However, in some cases a node that corresponds to a particular type will have unnamed node children that are not specified in the children spec.

Example:
```
  {
    "type": "block_expression",
    "named": true,
    "fields": {},
    "children": {
      "multiple": true,
      "required": false,
      "types": [
        {
          "type": "expression_statement",
          "named": true
        },
        {
          "type": "let_statement",
          "named": true
        },
        {
          "type": "return_statement",
          "named": true
        }
      ]
    }
  },
```

corresponding definition:
```
    block_expression: $ => seq('{', repeat($._statement), '}'),
```

iterating through the children of a node with kind "block_expression" will first yield a node with kind "{".

Something which you might not expect from the node-types.

Now a simple solution for the typed binding is to ignore unexpected unnamed nodes in sequences ( multiple: true ), and this is what I implemented.

However, it would be nice to either have the parser or the node-types explicitly indicate that such nodes are expected.

> I might just be doing everything wrong though 😅

**Comment by ahlinc** [comment]
For such extended cases the whole required info can be gathered from a `gammar.json` file that exists in an `src/` folder after a parser generation.
